botorch/profiling.py

```diff
@@ -0,0 +1,247 @@
+        87
+        88    185.1 MiB      0.0 MiB           1       test_X = torch.rand(batch_dim, q, dim)
+        89
+        92    185.1 MiB      0.0 MiB           1       wrapped_qnei_call = memory_profile(qNEI.__call__)
+        93    677.2 MiB    492.1 MiB           1       qNEI(test_X)
+        94                                             # wrapped_qnei_call(test_X)
+        95    699.0 MiB     21.9 MiB           1       qNEI(test_X)
+        96    721.3 MiB     22.2 MiB           1       qNEI(test_X)
+
+    Args:
+        q: Starts freezing up my Mac at 10000; q=1000 runs.
+        dim: dimension of feature space. Realistically might be 100.
+        n: number of points considered jointly. Realistically might be500
+        batch_dim: Default 5
+
+    """
+    train_X, train_Y = _get_train_x_and_y(n, dim, batch_dim)
+
+    model = SingleTaskGP(train_X, train_Y)
+    qNEI = qNoisyExpectedImprovement(model, X_baseline=train_X)
+
+    test_X = torch.rand(batch_dim, q, dim)
+
+    # we can't do `memory_profile(qNEI)` since `memory_profile` doesn't work on an
+    # instance, but profiling `qNEI.__call__` accomplishes the same thing.
+    # wrapped_qnei_call = memory_profile(qNEI.__call__)
+    qNEI(test_X)
+    qNEI(test_X)
+    qNEI(test_X)
+
+
+@memory_profile
+def run_qnehvi(q: int = 100, dim: int = 2, n: int = 50) -> None:
+    """
+    Noisy Expected Hypervolume Improvement
+
+    should have similar memory issues to run_qnei
+    Based on https://botorch.org/tutorials/multi_objective_bo
+    # TODO figure out what reasonable input dimensions are
+    """
+    tkwargs = {"device": _get_device(), "dtype": torch.double}
+    train_X = torch.rand(10, 4, **tkwargs)
+    train_Y = torch.sin(train_X[:, :1])
+    model = SaasFullyBayesianSingleTaskGP(
+        train_X=train_X, train_Y=train_Y, train_Yvar=None
+    )
+    # This is not the function we're interested in, but it needs to be run so
+    # the model can be fit.
+    fit_fully_bayesian_model_nuts(
+        model, warmup_steps=8, num_samples=5, thinning=2, disable_progbar=True
+    )
+    sampler = IIDNormalSampler(num_samples=2)
+    qNEHVI = qNoisyExpectedHypervolumeImprovement(  # noqa: F841
+        model=ModelListGP(model, model),
+        X_baseline=train_X,
+        ref_point=torch.zeros(2, **tkwargs),
+        sampler=sampler,
+    )
+    # TODO: figure out what to call qNEHVI on
+    raise NotImplementedError
+
+
+def run_fit_fully_bayesian_model_nuts(dim: int = 100, n: int = 200) -> None:
+    """
+    Profile `fit_fully_bayesian_model_nuts` for time.
+
+    * Almost all of the time in `fit_fully_bayesian_model_nuts` is spent in
+        `mcmc.run`.
+    * Almost all of the time in `mcmc.run` is spent in `MCMC.sampler.run`.
+    * Almost all of the time in `MCMC.sampler.run` is in `_gen_samples`
+    * The time in `_gen_samples` is mainly spent in `kernel.sample`, with some
+        spent in `kernel.setup`.
```

Comment:
Yeah, the vast majority of the time *should* (would be good verify) be spent here: https://github.com/pytorch/botorch/blob/main/botorch/models/fully_bayesian.py#L159-L168

This is computing the train-train covariance under the new parameter (lengthscale, outputscale) samples and then samples from a MVN with that covariance (which means an `O(n^3)` Cholesky decomposition where `n` is the number of training examples.